core/platform/src/error_code.rs

```diff
@@ -1,39 +1,232 @@
-/// A system call error code. This can either be an error code returned by the
-/// kernel or BADRVAL, which indicates the kernel returned the wrong type of
-/// response to a system call.
-// ErrorCode is not an enum so that conversion from the kernel's return value (a
-// `usize` in a register) is free.
+/// An error code returned by the kernel.
 // TODO: derive(Debug) is currently only enabled for test builds, which is
 // necessary so it can be used in assert_eq!. We should develop a lighter-weight
 // Debug implementation and see if it is small enough to enable on non-Debug
 // builds.
 #[cfg_attr(test, derive(Debug))]
 #[derive(Clone, Copy, PartialEq, Eq)]
-pub struct ErrorCode(usize);
+#[repr(u16)]  // To facilitate use with transmute() in CommandReturn
+#[rustfmt::skip]
+pub enum ErrorCode {
+    Fail = 1,
+    Busy = 2,
+    Already = 3,
+    Off = 4,
+    Reserve = 5,
+    Invalid = 6,
+    Size = 7,
+    Cancel = 8,
+    NoMem = 9,
+    NoSupport = 10,
+    NoDevice = 11,
+    Uninstalled = 12,
+    NoAck = 13,
 
-impl From<usize> for ErrorCode {
-    fn from(value: usize) -> ErrorCode {
-        ErrorCode(value)
-    }
+    // Error codes reserved for future use. We have to include these for future
+    // compatibility -- this allows process binaries compiled with this version
+    // of libtock-rs to run on future kernel versions that may return a larger
+    // variety of error codes.
```

Comment:
This does make sense, but I'm not sure whether it actually brings us any benefits: wouldn't a kernel, returning unexpected `ErrorCode` essentially always break the userspace apps because they wouldn't know how to deal with such an error? Or am I missing something here?

Reply:
Userspace code receiving "unexpected" error codes (error codes the system call's documentation says it will never return) is always a possibility. Code handling errors should always have an "I received an unexpected error" error handling path.

The only way we could avoid "unexpected errorrs" is to finalize the list of errors: change the TRD to say we can *never* add a new error type in Tock 2.x. Even then it would only change things for system calls that have documented reasons for returning all 13 error codes.